VMware: Inventory scenario guide for hostnames (#71055)
ansible/ansible: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_hostnames.rst

.. _vmware_ansible_inventory_using_hostnames:

*************************************************
Using VMware dynamic inventory plugin - Hostnames
*************************************************

.. contents:: Topics

> VMware Dynamic Inventory Plugin - Customizing hostnames

Reviewer: ```suggestion
VMware dynamic inventory plugin - Customizing hostnames
```

=======================================================

> VMware inventory plugin allows to configure hostnames using ``hostnames`` configuration parameter.

Reviewer: ```suggestion
VMware inventory plugin allows you to configure hostnames using the ``hostnames`` configuration parameter.
```

In this scenario guide we will see how you configure hostnames from the given VMware guest in the inventory.

Requirements
------------

To use the VMware dynamic inventory plugins, you must install `pyVmomi <https://github.com/vmware/pyvmomi>`_
on your control node (the host running Ansible).

> To include tag-related information for the virtual machines in your dynamic inventory, you also need the `vSphere Automation SDK <https://code.vmware.com/web/sdk/65/vsphere-automation-python>`_, which supports REST API features like tagging and content libraries, on your control node.

Reviewer: ```suggestion
To include tag-related information for the virtual machines in your dynamic inventory, you also need the `vSphere Automation SDK <https://code.vmware.com/web/sdk/65/vsphere-automation-python>`_, which supports REST API features such as tagging and content libraries, on your control node.
```

You can install the ``vSphere Automation SDK`` following `these instructions <https://github.com/vmware/vsphere-automation-sdk-python#installing-required-python-packages>`_.

.. code-block:: bash

    $ pip install pyvmomi

> VMware dynamic inventory plugin is now available in the VMware Collection, so you need to install VMware collection via -

Reviewer: ```suggestion
Starting in Ansible 2.10, the VMware dynamic inventory plugin is  available in the ``community.vmare`` collection included Ansible.
```

.. code-block:: bash

    $ ansible-galaxy collection install community.vmware

> 

Reviewer: since community.vmare is part of Ansible 2.10, you don't need the separate install above.  If you feel it's important to include this, then add a sentence above it to say "To install the latest ``community.vmare`` collection:"

> To use this VMware dynamic inventory plugin, you need to enable it first by specifying the following in the ``ansible.cfg`` file:

Reviewer: ```suggestion
To use this VMware dynamic inventory plugin:

 1.  Enable it first by specifying the following in the ``ansible.cfg`` file:
```

.. code-block:: ini

  [inventory]
  enable_plugins = community.vmware.vmware_vm_inventory

> Then, create a file that ends in ``.vmware.yml`` or ``.vmware.yaml`` in your working directory.

Reviewer: ```suggestion
2. Create a file that ends in ``.vmware.yml`` or ``.vmware.yaml`` in your working directory.
```

The ``vmware_vm_inventory`` inventory plugin takes in the same authentication information as any other VMware modules does.

Here's an example of a valid inventory file with custom hostname for the given VMware guest:

.. code-block:: yaml

    plugin: community.vmware.vmware_vm_inventory
    strict: False
    hostname: 10.65.223.31
    username: administrator@vsphere.local
    password: Esxi@123$%
    validate_certs: False
    with_tags: False
    hostnames:
    - config.name

> Here, we have configured custom hostname by specifying ``hostnames`` parameter to ``config.name``. This will reterive

Reviewer: ```suggestion
Here, we have configured a custom hostname by setting the ``hostnames`` parameter to ``config.name``. This will retrieve the
```

``config.name`` property from the virtual machine and populate it in the inventory.

> You can check all allowed properties for the given virtual machine, here - :ref:`vmware_inventory_vm_attributes`.

Reviewer: ```suggestion
You can check all allowed properties for the given virtual machine at :ref:`vmware_inventory_vm_attributes`.
```

Executing ``ansible-inventory --list -i <filename>.vmware.yml`` will create a list of the virtual machines that are ready to be configured using Ansible.

What to expect
--------------

> You will notice that instead of default behavior of representing hostname as ``config.name + _ + config.uuid``,

Reviewer: ```suggestion
You will notice that instead of default behavior of representing the hostname as ``config.name + _ + config.uuid``,
```

> inventory hosts show value as ``config.name``.

Reviewer: ```suggestion
the inventory hosts show value as ``config.name``.
```

.. code-block:: yaml

    {
      "_meta": {
        "hostvars": {
            "template_001": {
                "config.name": "template_001",
                "guest.toolsRunningStatus": "guestToolsNotRunning",
                ...
                "guest.toolsStatus": "toolsNotInstalled",
                "name": "template_001"
            },
            "vm_8046": {
                "config.name": "vm_8046",
                "guest.toolsRunningStatus": "guestToolsNotRunning",
                ...
                "guest.toolsStatus": "toolsNotInstalled",
                "name": "vm_8046"
            },
        ...
    }

Troubleshooting
---------------

> If custom property specified in ``hostnames`` fails -

Reviewer: ```suggestion
If the custom property specified in ``hostnames`` fails:
```

- Check if the values provided for username and password are correct.
- Check if you are using vCenter 6.5 and onwards to use this HTTP APIs.

> - Make sure it is a valid property, see :ref:`vmware_inventory_vm_attributes`

Reviewer: ```suggestion
- Make sure it is a valid property, see :ref:`vmware_inventory_vm_attributes`.
```
